pkg/controller/certificatesigningrequests/venafi/venafi.go

```diff
@@ -0,0 +1,206 @@
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	// The signing process with Venafi is slow. The "pickupID" allows us to track
+	// the progress of the certificate signing. It is set as an annotation the
+	// first time the Certificate is reconciled.
+	pickupID := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiPickupIDAnnotationKey]
+
+	// check if the pickup ID annotation is there, if not set it up.
+	if len(pickupID) == 0 {
+		pickupID, err := client.RequestCertificate(csr.Spec.Request, duration, customFields)
+		// Check some known error types
+		if err != nil {
+			switch err.(type) {
+
+			case venaficlient.ErrCustomFieldsType:
+				log.Error(err, "")
+				v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorCustomFields", err.Error())
+				util.CertificateSigningRequestSetFailed(csr, "ErrorCustomFields", err.Error())
+				_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+				return err
+
+			default:
+				message := fmt.Sprintf("Failed to request venafi certificate: %s", err)
+				log.Error(err, message)
+				v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorRequest", message)
+				util.CertificateSigningRequestSetFailed(csr, "ErrorRequest", message)
+				_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+				return err
+			}
+		}
+
+		if csr.Annotations == nil {
+			csr.Annotations = make(map[string]string)
+		}
+		csr.Annotations[experimentalapi.CertificateSigningRequestVenafiPickupIDAnnotationKey] = pickupID
+		_, err = v.certClient.Update(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	certPem, err := client.RetrieveCertificate(pickupID, csr.Spec.Request, duration, customFields)
+	if err != nil {
+		switch err.(type) {
+		case endpoint.ErrCertificatePending, endpoint.ErrRetrieveCertificateTimeout:
+			message := "Venafi certificate still in a pending state, waiting"
+			log.Error(err, message)
+			v.recorder.Event(csr, corev1.EventTypeNormal, "IssuancePending", message)
+			return err
+
+		default:
+			message := fmt.Sprintf("Failed to obtain venafi certificate: %s", err)
+			log.Error(err, message)
+			v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorRetrieve", message)
+			return err
+		}
+	}
+
+	bundle, err := utilpki.ParseSingleCertificateChainPEM(certPem)
+	if err != nil {
+		message := fmt.Sprintf("Failed to parse returned certificate bundle: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorParse", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorParse", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	csr.Status.Certificate = bundle.ChainPEM
+	csr, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+	if err != nil {
+		message := "Error updating certificate"
+		v.recorder.Eventf(csr, corev1.EventTypeWarning, "SigningError", "%s: %s", message, err)
+		return err
+	}
```

Comment:
question: why do we need to send an event here? We do not send events on optimistic locking errors in the rest of cert-manager, why would we here?

Reply:
There error might not be an optimistic locking error, it could be anything. The intention is to give as much info as possible _not the resource_ rather than just the logs.

I will just remove it if you think it is not necessary.

Reply:
On second thoughts.. It is important. Though should never _normally happen_, there is the case where cert-manager lacks the `sign` permission for this CSR. In this case, it would be important that this info is bubbled up to the user, rather than hidden in the logs.

Reply:
The case you mention wouldn't appear on the above UpdateStatus, right?

Reply:
It seems like there are two possible errors coming from this UpdateStatud:
- transient errors like network failures or optimistic locking,
- non-retriable errors like a missing sign permission.

We shouldn’t retry on the second case, it wouldn’t make sense to retry.

But for the sake of simplicity, let’s retry anyways.